httpx/client.py

```diff
@@ -131,171 +138,53 @@ def merge_headers(
         cert: CertTypes = None,
         timeout: TimeoutTypes = None,
     ) -> AsyncResponse:
-        if auth is None:
-            auth = self.auth
-
         url = request.url
-
         if url.scheme not in ("http", "https"):
             raise InvalidURL('URL scheme must be "http" or "https".')
 
-        if auth is None and (url.username or url.password):
-            auth = HTTPBasicAuth(username=url.username, password=url.password)
-
-        if auth is not None:
-            if isinstance(auth, tuple):
-                auth = HTTPBasicAuth(username=auth[0], password=auth[1])
-            request = auth(request)
-
-        response = await self.send_handling_redirects(
-            request,
-            verify=verify,
-            cert=cert,
-            timeout=timeout,
-            allow_redirects=allow_redirects,
-        )
-
-        if not stream:
-            try:
-                await response.read()
-            finally:
-                await response.close()
+        middlewares: typing.List[BaseMiddleware] = [
+            # TODO: RedirectMW uses a combination of parameters to the client
+            # but also to send, which makes this awkward instantiation
+            RedirectMiddleware(
+                allow_redirects=allow_redirects,
+                max_redirects=self.max_redirects,
+                base_cookies=self.cookies,
+            ),
+        ]
 
-        return response
+        if auth is None and (url.username or url.password):
+            middlewares.append(HTTPBasicAuthMiddleware(username=url.username, password=url.password))
+        elif isinstance(auth, tuple):
+            middlewares.append(HTTPBasicAuthMiddleware(username=auth[0], password=auth[1]))
+        elif isinstance(auth, HTTPBasicAuth):
+            middlewares.append(HTTPBasicAuthMiddleware(username=auth.username, password=auth.password))
 
-    async def send_handling_redirects(
-        self,
-        request: AsyncRequest,
-        *,
-        cert: CertTypes = None,
-        verify: VerifyTypes = None,
-        timeout: TimeoutTypes = None,
-        allow_redirects: bool = True,
-        history: typing.List[AsyncResponse] = None,
-    ) -> AsyncResponse:
-        if history is None:
-            history = []
+        middlewares.extend(self.additional_middlewares)
 
         while True:
-            # We perform these checks here, so that calls to `response.next()`
-            # will raise redirect errors if appropriate.
-            if len(history) > self.max_redirects:
-                raise TooManyRedirects()
-            if request.url in [response.url for response in history]:
-                raise RedirectLoop()
+            for middleware in middlewares:
```

Comment:
In order to implement caching we'd need a middleware to be able to return an HTTP response on the `process_request()` stage.

Reply:
Good point, is it worth adding it now?

Reply:
I think so, might as well get it working with this revision.